### PR TITLE
19529: add support for controller backup

### PR DESCRIPTION
**Pull request by shuoliugit**

### CODE REVIEW

File: goaviatrix/controller.go [hunk 1]
```diff
@@ -422,3 +430,54 @@ func (c *Client) SetControllerVpcDnsServer(enabled bool) error {
...
+	backupAccountName := d.Get("backup_account_name").(string)
+	backupBucketName := d.Get("backup_bucket_name").(string)
+	backupStorageName := d.Get("backup_storage_name").(string)
+	backupContainerName := d.Get("backup_container_name").(string)
+	backupRegion := d.Get("backup_region").(string)
+
+	if backupCloudType == 0 || backupAccountName == "" {
+		err := fmt.Errorf("please specify 'backup_cloud_type' and 'backup_account_name'" +
+			" to enable backup configuration")
+		return err
```
Comment (CyrusJavan) [inline goaviatrix/controller.go]: Why add the extra line to assign to a variable before returning? Does not seem more clear to me

File: goaviatrix/controller.go [hunk 2]
```diff
@@ -422,3 +430,54 @@ func (c *Client) SetControllerVpcDnsServer(enabled bool) error {
 		Action: action,
 	}, BasicCheck)
 }
+
+func ValidateBackupConfig(d *schema.ResourceData) error {
```
Comment (CyrusJavan) [inline goaviatrix/controller.go]: This function should not be in `goaviatrix` package. This is specific to the resource schema and should be in the controller config resource file.

File: docs/resources/aviatrix_controller_config.md [hunk 3]
```diff
@@ -69,11 +69,16 @@ The following arguments are supported:
...
 ### Backup
 * `backup_configuration` - (Optional) Switch to enable/disable controller CloudN backup config. Valid values: true, false. Default value: false.
-* `backup_cloud_type` - (Optional) Type of cloud service provider, requires an integer value. Use 1 for AWS.
+* `backup_cloud_type` - (Optional) Type of cloud service provider, requires an integer value. Use 1 for AWS, 4 for GCP, 8 for Azure, 16 for OCI, and 256 for AWSGOV.
 * `backup_account_name` - (Optional) Name of the cloud account in the Aviatrix controller.
-* `backup_bucket_name` - (Optional) S3 Bucket Name for AWS.
+* `backup_bucket_name` - (Optional) Bucket Name. Required for AWS, AWSGOV, GCP and OCI. 
+* `backup_storage_name` - (Optional) Storage name. Required for Azure.
+* `backup_container_name` - (Optional) Container name. Required for Azure.
+* `backup_region` - (Optional) Name of region. Required for Azure and OCI.
```
Comment (CyrusJavan) [inline docs/resources/aviatrix_controller_config.md]: Required **to enable configuration backup** for AWS, AWSGOV, GCP and OCI.

File: aviatrix/resource_aviatrix_controller_config.go [hunk 4]
```diff
@@ -468,77 +492,90 @@ func resourceAviatrixControllerConfigUpdate(d *schema.ResourceData, meta interfa
...
 	if d.HasChange("backup_configuration") {
-		backupConfiguration := d.Get("backup_configuration").(bool)
-		backupCloudType := d.Get("backup_cloud_type").(int)
-		backupAccountName := d.Get("backup_account_name").(string)
-		backupBucketName := d.Get("backup_bucket_name").(string)
-		multipleBackups := d.Get("multiple_backups").(bool)
 		if backupConfiguration {
-			if backupCloudType == 0 || backupAccountName == "" || backupBucketName == "" {
-				return fmt.Errorf("please specify 'backup_cloud_type', 'backup_account_name' and 'backup_bucket_name' to enable backup configuration")
+			backupConfigErr := goaviatrix.ValidateBackupConfig(d)
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_controller_config.go]: We should follow the pattern that is used throughout the code base and use the variable name `err` for error returns.

File: aviatrix/resource_aviatrix_controller_config.go [hunk 5]
```diff
@@ -212,30 +230,39 @@ func resourceAviatrixControllerConfigCreate(d *schema.ResourceData, meta interfa
...
+	backupStorageName := d.Get("backup_storage_name").(string)
+	backupContainerName := d.Get("backup_container_name").(string)
+	backupRegion := d.Get("backup_region").(string)
 	multipleBackups := d.Get("multiple_backups").(bool)
+
 	if backupConfiguration {
-		if backupCloudType == 0 || backupAccountName == "" || backupBucketName == "" {
-			return fmt.Errorf("please specify 'backup_cloud_type', 'backup_account_name' and 'backup_bucket_name'" +
-				" to enable backup configuration")
+		backupConfigErr := goaviatrix.ValidateBackupConfig(d)
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_controller_config.go]: We should follow the pattern that is used throughout the code base and use the variable name `err` for error returns.

**Review by CyrusJavan — CHANGES_REQUESTED** [review]

**Review by CyrusJavan — APPROVED** [review]

**Review by zhiqiangjin — APPROVED** [review]

**Pull request merged**